src/VisualStudio/Core/Def/Implementation/Workspace/VisualStudioActiveDocumentTracker.cs

```diff
@@ -0,0 +1,290 @@
+        /// Listens to frame notifications for a visible frame. When the frame becomes invisible or closes,
+        /// then it automatically disconnects.
+        /// </summary>
+        private class FrameListener : IVsWindowFrameNotify, IVsWindowFrameNotify2
+        {
+            public readonly IVsWindowFrame Frame;
+
+            private readonly VisualStudioActiveDocumentTracker _documentTracker;
+            private readonly uint _frameEventsCookie;
+
+            private readonly ITextBuffer _textBuffer;
+
+            public FrameListener(VisualStudioActiveDocumentTracker service, IVsWindowFrame frame)
+            {
+                _documentTracker = service;
+                _documentTracker.AssertIsForeground();
+
+                this.Frame = frame;
+
+                ((IVsWindowFrame2)frame).Advise(this, out _frameEventsCookie);
+
+                if (ErrorHandler.Succeeded(frame.GetProperty((int)__VSFPROPID.VSFPROPID_DocData, out var docData)))
+                {
+                    if (docData is IVsTextBuffer bufferAdapter)
+                    {
+                        _textBuffer = _documentTracker._editorAdaptersFactoryService.GetDocumentBuffer(bufferAdapter);
+
+                        if (!_textBuffer.ContentType.IsOfType(ContentTypeNames.RoslynContentType))
+                        {
+                            _textBuffer.Changed += NonRoslynTextBuffer_Changed;
+                        }
+                    }
+                }
+            }
+
+            private void NonRoslynTextBuffer_Changed(object sender, TextContentChangedEventArgs e)
+            {
+                _documentTracker.DocumentsChanged?.Invoke(_documentTracker, EventArgs.Empty);
+            }
+
+            /// <summary>
+            /// Returns the current DocumentId for this window frame. Care must be made with this value, since "current" could change asynchronously as the document
+            /// could be unregistered from a workspace.
+            /// </summary>
+            public DocumentId GetDocumentId(Workspace workspace)
+            {
+                if (_textBuffer == null)
+                {
+                    return null;
+                }
+
+                var textContainer = _textBuffer.AsTextContainer();
+                return workspace.GetDocumentIdInCurrentContext(textContainer);
+            }
+
+            int IVsWindowFrameNotify.OnDockableChange(int fDockable)
+            {
+                return VSConstants.S_OK;
+            }
+
+            int IVsWindowFrameNotify.OnMove()
+            {
+                return VSConstants.S_OK;
+            }
+
+            int IVsWindowFrameNotify.OnShow(int fShow)
+            {
+                switch ((__FRAMESHOW)fShow)
+                {
+                    case __FRAMESHOW.FRAMESHOW_WinClosed:
+                    case __FRAMESHOW.FRAMESHOW_WinHidden:
+                    case __FRAMESHOW.FRAMESHOW_TabDeactivated:
+                        return Disconnect();
```

Comment:
i don't know enough about these events.  is it possible we could disconnect, but then the doc gets visible again somehow?

Reply:
Correct: if the window is hidden (say you switch to another tab), we disconnect. When the frame recieves focus again, our "active frame changed" code runs, and we resubscribe again.